js_terminal.nim

```diff
@@ -0,0 +1,341 @@
+      result &= " text-decoration: underline line-through;"
+    else:
+      result &= " text-decoration: underline;"
+  elif styleStrikethrough in st:
+    result &= " text-decoration: line-through;"
+  #
+  result &= "\""  # 'style=' ends
+  #
+  # blinking classes
+  #
+  if styleBlink in st:
+    if styleBlinkRapid in st:
+      result &= " class=\"blink blinkrapid\""
+    else:
+      result &= " class=\"blink\""
+  elif styleBlinkRapid in st:
+    result &= " class=\"blinkrapid\""
+  #
+  result &= ">"
+
+
+proc genRowHtml(tb: TerminalBuffer, row: int): string =
+  var fg = tb[0, row].fg
+  var bg = tb[0, row].bg
+  var st = tb[0, row].style
+  result = genStartSpan(fg, bg, st)
+  for col in 0 ..< tb.width:
+    let c = tb[col, row]
+    if (c.fg != fg) or (c.bg != bg) or (c.style != st):
+      result &= "</span>"
+      result &= genStartSpan(c.fg, c.bg, c.style)
+      fg = c.fg
+      bg = c.bg
+      st = c.style
+    if cursorBlockVisible and (row == cursorY) and (col == cursorX):
+      result &= "█"
+    else:
+      result &= $c.ch
+  result &= "</span>"
+
+
+proc applyTbToPre(tb: TerminalBuffer) =
+  if preTag.isNil:
+    # echo "(attp) not loaded yet"
+    return
+  setCursorPos(tb.currX, tb.currY)
+  var newHtml = ""
+  for row in 0 ..< tb.height:
+    newHtml &= tb.genRowHtml(row)
+    if row < (tb.height - 1):
+      newHtml &= "\n"
+  preTag.innerHtml = newHtml.cstring
+
+
+proc eraseScreen*() =
+  discard
+
+
+proc put*(tb: TerminalBuffer) = 
+  applyTbToPre(tb)
+
+
+proc setControlCHook*(hook: proc () {.noconv.}) =
+  discard
+
+
+proc unsetControlCHook*() =
+  discard
+
+
+proc onkeydown(ev: Event) =
+  # echo $ev.`type`
+  var temp = cast[KeyboardEvent](ev)
+  lastKey = temp.key
+
+
+let jsKeyMap = {
+  "ArrowUp": Key.Up,
+  "ArrowDown": Key.Down,
+  "ArrowLeft": Key.Left,
+  "ArrowRight": Key.Right
+}.toTable
+
+proc getKeyAsync*(): Key =
+  result = Key.None
+  let rawKey = $lastKey
+  lastKey = "".cstring
+  case len(rawKey):
+  of 0:
+    discard
+  of 1:
+    var asciiInt = rawKey[0].ord
+    result = asciiInt.toKey()
+  else:
+    for keyItem in Key:
+      if $keyItem == rawKey:
+        result = keyItem
+        break
+    if result == Key.None:
+      if jsKeyMap.hasKey(rawKey):
+        result = jsKeyMap[rawKey]
+
+
+proc dummyInterval() = 
+  discard
+
+
+proc cursorBlinker() =
+  if cursorOn:
+    cursorBlockVisible = not cursorBlockVisible
+  else:
+    cursorBlockVisible = false
+
+
+proc consoleInit*() =
+  window.addEventListener(
+    "keydown".cstring,
+    onkeydown,
+    false
+  )
+  globalTimer = window.setInterval(dummyInterval, 6000)
+  globalTimer = window.setInterval(cursorBlinker, 800)
+  if not preTag.isNil:
+    preTag.disabled = false
+
+
+proc consoleDeInit*() =
+  window.clearInterval(globalTimer)
+  window.clearInterval(cursorTimer)
+  window.removeEventListener(
+    "keydown".cstring, 
+    onkeydown
+  )
+  if preTag.isNil:
+    echo "cdi not loaded yet"
```

Comment:
Can we provide a more useful message? What is `cdi`, is the user supposed to know that?